### PR TITLE
allow to select only scenarios with pvfs on the smaa tab

### PR DESCRIPTION
**Pull request by kzalite**

### CODE REVIEW

File: app/ts/PreferencesTab/PreferencesUtil.test.ts [hunk 1]
```diff
@@ -166,4 +171,48 @@ describe('PreferencesUtil', () => {
       expect(result).toEqual(expectedResult);
     });
   });
+
+  describe('filterScenariosWithPvfs', () => {
+    it('should return scenarios that have datasources and set pvs', () => {
```
Comment (DanielReid) [inline app/ts/PreferencesTab/PreferencesUtil.test.ts:176]: pvfs

File: app/ts/PreferencesTab/PreferencesUtil.test.ts [hunk 2]
```diff
@@ -166,4 +171,48 @@ describe('PreferencesUtil', () => {
...
+              }
+            }
+          },
+          subproblemId: '37',
+          workspaceId: '42'
+        }
+      };
+      const result = filterScenariosWithPvfs(scenarios);
+      const expectedResult = {
+        scenarioId1: scenarios['scenarioId1']
```
Comment (DanielReid) [inline app/ts/PreferencesTab/PreferencesUtil.test.ts:213]: `scenarios.scenarioId1`

File: app/ts/PreferencesTab/PreferencesUtil.ts [hunk 3]
```diff
@@ -67,3 +68,23 @@ export function buildScenarioWithPreferences(
...
+export function filterScenariosWithPvfs(
+  scenarios: Record<string, IMcdaScenario>
+): Record<string, IMcdaScenario> {
+  return _(scenarios)
+    .filter((scenario: IMcdaScenario) => {
+      return _.every(
+        scenario.state.problem.criteria,
+        (criterion: IScenarioCriterion) => {
+          return (
+            !!criterion.dataSources &&
```
Comment (DanielReid) [inline app/ts/PreferencesTab/PreferencesUtil.ts:81]: what's the double `!` preventing?

File: app/ts/PreferencesTab/PreferencesUtil.ts [hunk 4]
```diff
@@ -67,3 +68,23 @@ export function buildScenarioWithPreferences(
...
+            !!criterion.dataSources[0].pvf.direction &&
+            !!criterion.dataSources[0].pvf.type
+          );
+        }
+      );
+    })
+    .keyBy('id')
+    .value();
```
Comment (chinook25) [inline app/ts/PreferencesTab/PreferencesUtil.ts:89]: Could this be _.omit when then don't have the properties?

File: app/ts/PreferencesTab/PreferencesUtil.ts [hunk 5]
```diff
@@ -67,3 +68,23 @@ export function buildScenarioWithPreferences(
   };
   return {..._.omit(scenario, ['state']), state: newState};
 }
+
+export function filterScenariosWithPvfs(
+  scenarios: Record<string, IMcdaScenario>
+): Record<string, IMcdaScenario> {
+  return _(scenarios)
+    .filter((scenario: IMcdaScenario) => {
```
Comment (DanielReid) [inline app/ts/PreferencesTab/PreferencesUtil.ts:76]: `filter -> keyBy` is equivalent to `omit/omitBy` afaik

**Review by DanielReid — APPROVED** [review]
finel; minor niggles

**Review by chinook25 — APPROVED** [review]

File: app/ts/PreferencesTab/PreferencesUtil.ts [hunk 6]
```diff
@@ -67,3 +68,23 @@ export function buildScenarioWithPreferences(
...
+export function filterScenariosWithPvfs(
+  scenarios: Record<string, IMcdaScenario>,
+  problemCriteria: Record<string, IProblemCriterion>
+): Record<string, IMcdaScenario> {
+  return _.pickBy(scenarios, (scenario: IMcdaScenario) =>
+    _.every(
+      scenario.state.problem.criteria,
+      (criterion: IScenarioCriterion, criterionId: string) => {
+        const scenarioPvf = getScenarioPvf(criterionId, scenario);
+        const problemPvf = getProblemPvf(criterionId, problemCriteria);
```
Comment (chinook25) [inline app/ts/PreferencesTab/PreferencesUtil.ts:81]: We should really start splitting out these things on upload :'( All the special casing :(

**Review by chinook25 — APPROVED** [review]

**Pull request closed without merging**